File: src/main/java/me/deecaad/weaponmechanicsplus/weapon/modifiers/ScopeModifier.java
package me.deecaad.weaponmechanicsplus.weapon.modifiers;

import me.deecaad.core.file.SerializeData;
import me.deecaad.core.file.Serializer;
import me.deecaad.core.file.SerializerException;
import me.deecaad.core.utils.EnumUtil;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.DoubleModifier;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.Operation;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class ScopeModifier implements Serializer<ScopeModifier> {

    private DoubleModifier zoomAmount;
    private Optional<Boolean> isNightVision;  // optional for 3 states, isEmpty -> no mod, true -> add night vision, false -> remove night vision
    private List<DoubleModifier> zoomStacking;

    public ScopeModifier() {
    }

    public ScopeModifier(DoubleModifier zoomAmount, Boolean isNightVision, List<DoubleModifier> zoomStacking) {
        this.zoomAmount = zoomAmount == null ? DoubleModifier.NONE : zoomAmount;
        this.isNightVision = Optional.ofNullable(isNightVision);
        this.zoomStacking = zoomStacking;
    }

    public DoubleModifier getZoomAmount() {
        return zoomAmount;
    }

    public Optional<Boolean> getIsNightVision() {
        return isNightVision;
    }

    public List<DoubleModifier> getZoomStacking() {
        return zoomStacking;
    }

    @NotNull
    @Override
    public ScopeModifier serialize(SerializeData data) throws SerializerException {
        DoubleModifier zoomAmount = data.of("Zoom_Amount").serialize(DoubleModifier.class);
        Boolean isNightVision = data.has("Night_Vision") ? data.of("Night_Vision").assertExists().getBool() : null;

        List<String[]> splits = data.ofList("Zoom_Stacking")
                .addArgument(Operation.class, true)
                .addArgument(double.class, true)
                .assertList().get();

        List<DoubleModifier> zoomStacking = new ArrayList<>();
        for (String[] split : splits) {
            zoomStacking.add(new DoubleModifier(EnumUtil.getIfPresent(Operation.class, split[0]).orElseThrow(), Double.parseDouble(split[1])));
        }

        return new ScopeModifier(zoomAmount, isNightVision, zoomStacking);
    }
}

File: src/main/kotlin/me/deecaad/weaponmechanicsplus/weapon/modifiers/ScopeModifier.kt
package me.deecaad.weaponmechanicsplus.weapon.modifiers

import me.deecaad.core.file.*
import me.deecaad.core.utils.EnumUtil
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.*
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.MechanicsModifier.Companion.serializeMechanicsModifier
import java.util.*

class ScopeModifier : Serializer<ScopeModifier> {

    var zoomAmount: DoubleModifier? = null
    var isNightVision: Boolean? = null
    var zoomStacking: List<DoubleModifier?> = listOf()
    var mechanicsModifier: MechanicsModifier? = null

    /**
     * Default constructor for serializer
     */
    constructor()

    constructor(zoomAmount: DoubleModifier?, isNightVision: Boolean?, zoomStacking: List<DoubleModifier?>,
                mechanicsModifier: MechanicsModifier?) {
        this.zoomAmount = zoomAmount
        this.isNightVision = isNightVision
        this.zoomStacking = zoomStacking
        this.mechanicsModifier = mechanicsModifier
    }

    @Throws(SerializerException::class)
    override fun serialize(data: SerializeData): ScopeModifier {
        val zoomAmount = data.of("Zoom_Amount").serialize(DoubleModifier::class.java)
        val isNightVision = if (data.has("Night_Vision")) data.of("Night_Vision").assertExists().bool else null

        val splits = data.ofList("Zoom_Stacking")
            .addArgument(Operation::class.java, true)
            .addArgument(Double::class.javaPrimitiveType, true)
            .assertList().get()

        val zoomStacking: MutableList<DoubleModifier> = ArrayList()
        for (split in splits) {
            val math = EnumUtil.getIfPresent(Operation::class.java, split[0]).orElseThrow()
            zoomStacking.add(DoubleModifier(math, split[1].toDouble()))
        }

        val mechanicsModifier = data.serializeMechanicsModifier()

        return ScopeModifier(zoomAmount, isNightVision, zoomStacking, mechanicsModifier)
    }
}